frame/support/src/traits/tokens/fungibles.rs

```diff
@@ -126,18 +151,54 @@ pub trait Reserve<AccountId>: Inspect<AccountId> {
 
 	/// Check to see if some `amount` of `asset` may be reserved on the account of `who`.
 	fn can_reserve(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> bool;
+}
 
+/// Trait for mutating a set of named fungible assets which can be reserved.
+pub trait MutateReserve<AccountId>: InspectReserve<AccountId> + Transfer<AccountId> {
 	/// Reserve some funds in an account.
 	fn reserve(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> DispatchResult;
 
 	/// Unreserve some funds in an account.
-	fn unreserve(asset: Self::AssetId, who: &AccountId, amount: Self::Balance) -> DispatchResult;
+	fn unreserve(asset: Self::AssetId, who: &AccountId, amount: Self::Balance)
+		-> Result<Self::Balance, DispatchError>;
 
-	/// Transfer reserved funds into another account.
+	/// Transfer reserved funds into another account. Done on a best-effort basis.
 	fn repatriate_reserved(
 		asset: Self::AssetId,
-		who: &AccountId,
+		source: &AccountId,
+		dest: &AccountId,
 		amount: Self::Balance,
-		status: BalanceStatus,
-	) -> DispatchResult;
+	) -> Result<Self::Balance, DispatchError> {
+		// Done on a best-effort basis. Basically just a slash + transfer
+		let actual = Self::unreserve(asset, source, amount)?;
+		let actual = <Self as fungibles::Transfer<AccountId>>::transfer(asset, source, dest, actual)?;
+		Ok(actual)
+	}
+}
+
+/// Trait for mutating one of several types of fungible assets which can be reserved.
+pub trait BalancedReserve<AccountId>: Balanced<AccountId> + MutateReserve<AccountId> {
```

Comment:
move to `balanced.rs`?

Reply:
I'd say it probably belongs just as much with the other `Reserve` traits...